xonsh/lexer.py

```diff
@@ -1,13 +1,307 @@
+        '>>': 'RSHIFT', '<': 'LT', '<=': 'LE', '>': 'GT', '>=': 'GE',
+        '==': 'EQ', '!=': 'NE', '->': 'RARROW',
+        # assignment operators
+        '=': 'EQUALS', '+=': 'PLUSEQUAL', '-=': 'MINUSEQUAL',
+        '*=': 'TIMESEQUAL', '/=': 'DIVEQUAL', '%=': 'MODEQUAL',
+        '**=': 'POWEQUAL', '<<=': 'LSHIFTEQUAL', '>>=': 'RSHIFTEQUAL',
+        '&=': 'AMPERSANDEQUAL', '^=': 'XOREQUAL', '|=': 'PIPEEQUAL',
+        '//=': 'DOUBLEDIVEQUAL',
+}
+for (op, type) in op_map.items():
+    token_map[(tokenize.OP, op)] = type
+
+token_map[tokenize.NAME] = 'NAME'
+token_map[tokenize.NUMBER] = 'NUMBER'
+token_map[tokenize.STRING] = 'STRING'
+
+
+def handle_indent(state, token, stream):
+    level = len(token.string)
+    state['last'] = token
+    if token.type == tokenize.DEDENT:
+        state['indents'].pop()
+        yield _new_token('DEDENT', ' '*state['indents'][-1], token.start)
+    elif token.type == tokenize.INDENT:
+        state['indents'].append(level)
+        yield _new_token('INDENT', token.string, token.start)
+
+    try:
+        n = next(stream)
+    except:
+        n = None
+    if n is not None:
+        if n.type != tokenize.ENDMARKER:
+            for i in handle_token(state, n, stream):
+                yield i
+
+
+def handle_dollar(state, token, stream):
+    try:
+        n = next(stream)
+    except:
+        m = "missing token after $"
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+    if n.start != token.end:
+        m = "unexpected whitespace after $"
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+    if n.type == tokenize.NAME:
+        state['last'] = n
+        yield _new_token('DOLLAR_NAME', '$' + n.string, token.start)
+    elif n.type == tokenize.OP and n.string == '(':
+        state['pymode'].append(False)
+        state['last'] = n
+        yield _new_token('DOLLAR_LPAREN', '$(', token.start)
+    elif n.type == tokenize.OP and n.string == '[':
+        state['pymode'].append(False)
+        state['last'] = n
+        yield _new_token('DOLLAR_LBRACKET', '$[', token.start)
+    elif n.type == tokenize.OP and n.string == '{':
+        state['pymode'].append(True)
+        state['last'] = n
+        yield _new_token('DOLLAR_LBRACE', '${', token.start)
+    else:
+        e = 'expected NAME, (, [, or {{ after $, but got {0}'
+        m = e.format(n)
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+
+def handle_at(state, token, stream):
+    try:
+        n = next(stream)
+    except:
+        m = "missing token after @"
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+    if n.type == tokenize.OP and n.string == '(' and \
+            n.start == token.end:
+        state['pymode'].append(True)
+        yield _new_token('AT_LPAREN', '@(', token.start)
+        state['last'] = n
+    else:
+        yield _new_token('AT', '@', token.start)
+        state['last'] = token
+        for i in handle_token(state, n, stream):
+            yield i
```

Comment:
Unless we are trying explicitly for Python 2 compatibility here, you can use `yield from` here instead of the loop:

```python
yield from handle_token(state, n, stream)
```

This comment applies to all other similar instances in the file.